L1Trigger/L1CaloTrigger/plugins/Phase2L1CaloPFClusterEmulator.cc

```diff
@@ -0,0 +1,266 @@
+#include "FWCore/Utilities/interface/StreamID.h"
+
+#include "DataFormats/L1TCalorimeterPhase2/interface/CaloCrystalCluster.h"
+#include "DataFormats/L1TCalorimeterPhase2/interface/CaloTower.h"
+#include "DataFormats/L1TCalorimeterPhase2/interface/CaloPFCluster.h"
+#include "DataFormats/L1Trigger/interface/EGamma.h"
+
+#include <ap_int.h>
+#include <fstream>
+#include <iomanip>
+#include <iostream>
+#include <stdio.h>
+#include "Phase2L1CaloPFClusterEmulator.h"
+
+//#include "Phase2L1RCT.h"
+//#include "Phase2L1GCT.h"
+//#include "Phase2L1GCT_algo.h"
+
+
+//
+// class declaration
+//
+
+class Phase2L1CaloPFClusterEmulator : public edm::stream::EDProducer<> {
+public:
+  explicit Phase2L1CaloPFClusterEmulator(const edm::ParameterSet&);
+  ~Phase2L1CaloPFClusterEmulator();
+
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+private:
+  void beginStream(edm::StreamID) override;
+  void produce(edm::Event&, const edm::EventSetup&) override;
+  void endStream() override;
+
+  // ----------member data ---------------------------
+  edm::EDGetTokenT<l1tp2::CaloTowerCollection> caloTowerToken_;
+};
+
+//
+// constants, enums and typedefs
+//
+
+
+//
+// static data member definitions
+//
+
+//
+// constructors and destructor
+//
+Phase2L1CaloPFClusterEmulator::Phase2L1CaloPFClusterEmulator(const edm::ParameterSet& iConfig) :
+  caloTowerToken_(consumes<l1tp2::CaloTowerCollection>( iConfig.getParameter<edm::InputTag>( "gctFullTowers")))
+{
+  produces<l1tp2::CaloPFClusterCollection>( "GCTPFCluster" ) ;
+}
+
+Phase2L1CaloPFClusterEmulator::~Phase2L1CaloPFClusterEmulator() {
+}
+
+//
+// member functions
+//
+
+// ------------ method called to produce the data  ------------
+void Phase2L1CaloPFClusterEmulator::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  using namespace edm;
+  std::unique_ptr<l1tp2::CaloPFClusterCollection> pfclusterCands(new l1tp2::CaloPFClusterCollection);
+
+  edm::Handle<std::vector<l1tp2::CaloTower>> caloTowerCollection;
+  if(!iEvent.getByToken(caloTowerToken_, caloTowerCollection)) edm::LogError("Phase2L1CaloPFClusterEmulator") << "Failed to get towers from caloTowerCollection!" ;
+
+  iEvent.getByToken(caloTowerToken_, caloTowerCollection);
+  float GCTintTowers[34][72];
+  float realEta[34][72];
+  float realPhi[34][72];
+  for (const l1tp2::CaloTower &i : *caloTowerCollection) {
+    //std::cout << "Phase2L1CaloPFClusterEmulator: GCT FULL tower energy " << i.ecalTowerEt()
+    //          << " at eta, phi: (" << i.towerIEta() << ", " << i.towerIPhi() << ")" << std::endl; 
+
+    int ieta = i.towerIEta(); int iphi = i.towerIPhi();
+    GCTintTowers[ieta][iphi] = i.ecalTowerEt();
+    realEta[ieta][iphi] = i.towerEta();
+    realPhi[ieta][iphi] = i.towerPhi();
+
+  }
+
+  float regions[36][21][8];
+
+  for(int ieta = 0; ieta < 21; ieta++){
+    for(int iphi = 0; iphi < 8; iphi++){
+      for(int k = 0; k < 36; k++){
+        regions[k][ieta][iphi] = 0.;
```

Comment:
Similarly, for `21`, `8` and `36`, having constants to clear up the meaning here might be helpful